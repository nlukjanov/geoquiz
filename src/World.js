import React, { useState, useEffect } from 'react';
import axios from 'axios';

const World = () => {
  const [data, setData] = useState([]);
  const [question, setQuestion] = useState({
    flag: '',
    correctAnswer: '',
    answers: []
  });

  const [score, setScore] = useState(0);
  const [guess, setGuess] = useState('');

  const getData = async () => {
    try {
      const response = await axios.get('https://restcountries.eu/rest/v2/all');
      setData(response.data);
    } catch (error) {
      console.log(error);
    }
  };

  useEffect(() => {
    getData();
  }, []);

  useEffect(() => {
    if (data.length !== 0) {
      makeQuestion();
    }
  }, [data]);

  const makeQuestion = () => {
    const questionCountry = data[Math.floor(Math.random() * data.length)];
    const flag = questionCountry.flag;
    const correctAnswer = questionCountry.name;
    const excludeQuestionCountry = data.filter((country) => {
      return country !== questionCountry;
    });
    const getWrongAnswers = (sourceArray, neededElements) => {
      const result = [];
      for (let i = 0; i < neededElements; i++) {
        const index = Math.floor(Math.random() * sourceArray.length);
        result.push(sourceArray[index].name);
        sourceArray.splice(index, 1);
      }
      return result;
    };
    const answers = [
      ...getWrongAnswers(excludeQuestionCountry, 3),
      correctAnswer
    ];
    setQuestion({
      flag,
      correctAnswer,
      answers: shuffleArray(answers)
    });
  };

  const shuffleArray = (array) => {
    const newArray = [];
    while (array.length) {
      const randomIndex = Math.floor(Math.random() * array.length),
        element = array.splice(randomIndex, 1);
      newArray.push(element[0]);
    }
    return newArray;
  };

  const handleAnswerClick = (e) => {
    console.log(e.target.textContent);
    if (e.target.textContent === question.correctAnswer) {
      setScore(score + 1);
      setGuess('Correct');
    } else {
      setGuess('Incorrect');
    }
  };

  return (
    <div>
      <h2 data-testid='h2 title'>Guess the flag</h2>
      <img
        data-testid='flag image'
        src={question.flag}
        style={{ width: '400px', height: '200px' }}
        alt='flag'
      />
      <div data-testid='answers block'>
        {question.answers.map((answer, index) => (
          <button data-testid={`button ${index}`} onClick={handleAnswerClick} key={index}>
            {answer}
          </button>
        ))}
      </div>
      <div>{guess && <div>You are {guess}</div>}</div>
    </div>
  );
};

export default World;
