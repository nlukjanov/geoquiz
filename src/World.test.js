import React from 'react';
import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import '@testing-library/jest-dom/extend-expect';

import World from './World';

describe('World component', () => {

  const question = {
    flag: '',
    correctAnswer: '',
    answers: []
  };

  it('should render an h2 title', () => {
    const { getByTestId } = render(
      <BrowserRouter>
        <World />
      </BrowserRouter>
    );
    expect(getByTestId(/h2 title/i)).toBeTruthy;
  });

  it('should render an img tag with flag', () => {
    const { getByTestId } = render(
      <BrowserRouter>
        <World />
      </BrowserRouter>
    );
    expect(getByTestId(/flag image/i)).toBeTruthy;
  });

  it('should render an img tag with flag', () => {
    const { getByTestId } = render(
      <BrowserRouter>
        <World />
      </BrowserRouter>
    );
    expect(getByTestId(/flag image/i)).toBeTruthy;
  });

  it('should render a div with answers block', () => {
    const { getByTestId } = render(
      <BrowserRouter>
        <World />
      </BrowserRouter>
    );
    expect(getByTestId(/answers block/i)).toBeTruthy;
  });

  it('should render answer buttons', () => {
    const { getByTestId } = render(
      <BrowserRouter>
        <World />
      </BrowserRouter>
    );
    expect(getByTestId(/button 1/i)).toBeTruthy;
    expect(getByTestId(/button 2/i)).toBeTruthy;
    expect(getByTestId(/button 3/i)).toBeTruthy;
    expect(getByTestId(/button 4/i)).toBeTruthy;
  });
});
