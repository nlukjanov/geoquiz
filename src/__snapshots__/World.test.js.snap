// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`World component should render an h2 title 1`] = `
<div>
  <h2
    data-testid="h2 title"
  >
    Guess the flag
  </h2>
  <img
    alt="flag"
    src=""
    style="width: 400px; height: 200px;"
  />
  <div />
  <div />
</div>
`;
